### PR TITLE
Rose fileinstall recording.convert functional tests

### LINKED ISSUE #12: functional tests
Test the cylc-rose functionality from within the cylc-rose repository (rather than in cylc-flow).

There are reasons why this should be done here:

* Nicer (less stuff in cylc-flow).
* Ensures we have the correct Cylc/Rose version bracketing.
* Allows us to have a testing matrix with different Rose/Cylc versions to ensure inter-compatibility.

Suggested approach, use pytest, avoid the filesystem where possible:

Split the existing functions into two to separate the cylc interface and fileparsing logic from the cylc-rose internals e.g:

```python
# the entry point
def get_rose_vars(dir_, opts):
    # illustrative and very non-functional :/
    config_tree = rose_config_tree_loader(dir_, opts)
    node = config_tree.node
    vars = _get_rose_vars(node)
    # ...
    config = {}
    config['whatever'] = vars['whatever']
    # ...
    return {'whatever': vars}

def _get_rose_vars(node):
    # ...
```

Then test the internal cylc-rose bits without the cylc-flow interface overheads and associated filesystem meddling:

```python
def test_env():
    # illustrative and very non-functional :/
    node = ConfigNode()
    node.set(['env', 'foo'], '1')
    assert _get_rose_vars(node) == {
        'whatever': {
             'foo' 1
        }
    }
```

### PR DESCRIPTION
**Pull request by wxtim**
Closes #12 within this repo. A companion Cylc PR is still required.

Move functional tests for cylc-rose currently in the Cylc-flow repo's prove based framework to using Pytest in this repo.

### CODE REVIEW

File: tests/functional/04_opts_set_from_env/04-opts-set-from-env/flow.cylc [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!jinja2
```
Comment (oliver-sanders) [inline tests/functional/04_opts_set_from_env/04-opts-set-from-env/flow.cylc:1]: Is this meant to have `04_opts_set_from_env` in the path twice?

(same for some other files)

File: tests/functional/test_pre_configure.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    sub = subprocess.run(
+        ['cylc', 'validate', str(srcdir)], capture_output=True
+    )
+
+    assert sub.returncode != 0
+    if expect:
+        assert sub.stderr == (
+                "FileParseError: Plugins set templating engine = empy which "
+                "does not match #!jinja2 set in flow.cylc."
+            )
```
Comment (oliver-sanders) [inline tests/functional/test_pre_configure.py:48]: Unused?

**Review by oliver-sanders — COMMENTED** [review]
Tests seem unhappy.

File: tests/functional/04_opts_set_from_env/04-opts-set-from-env/flow.cylc (hunk 1 above)
Comment (wxtim) [reply tests/functional/04_opts_set_from_env/04-opts-set-from-env/flow.cylc:1]: No. This is a mistake.

Comment (wxtim) [reply tests/functional/04_opts_set_from_env/04-opts-set-from-env/flow.cylc:1]: Can you see this when browsing the source? I can't, either on GH or locally!

File: tests/functional/test_pre_configure.py (hunk 2 above)
Comment (wxtim) [reply tests/functional/test_pre_configure.py]: It is. It shouldn't be.

**Review by oliver-sanders — APPROVED** [review]

**Review by datamel — APPROVED** [review]

**Pull request merged**